FileSystem::openFile() should return a FileHandle

cdumez opened the pull request:
#### cb506c8dfb5d786161414aa588f9286266b60aa5
<pre>
FileSystem::openFile() should return a FileHandle
<a href="https://bugs.webkit.org/show_bug.cgi?id=289509">https://bugs.webkit.org/show_bug.cgi?id=289509</a>

Reviewed by Geoffrey Garen.

FileSystem::openFile() should return a FileHandle instead of a PlatformFileHandle, FileHandle being a
RAII wrapper to make sure we never forget to close file descriptors.

In a follow-up, I will look into cleaning up FileSystem.h, now that some of the API is only needed
by FileHandle.

* Source/JavaScriptCore/API/JSScript.mm:
(-[JSScript readCache]):
* Source/JavaScriptCore/jsc.cpp:
(JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/wasm/WasmStreamingParser.cpp:
(JSC::Wasm::dumpWasmSource):
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/FileHandle.cpp: Added.
(WTF::FileSystemImpl::FileHandle::FileHandle):
(WTF::FileSystemImpl::FileHandle::~FileHandle):
(WTF::FileSystemImpl::FileHandle::operator=):
(WTF::FileSystemImpl::FileHandle::readAll):
(WTF::FileSystemImpl::FileHandle::read):
(WTF::FileSystemImpl::FileHandle::write):
(WTF::FileSystemImpl::FileHandle::lock):
(WTF::FileSystemImpl::FileHandle::truncate):
(WTF::FileSystemImpl::FileHandle::closeIfNecessary):
(WTF::FileSystemImpl::FileHandle::size):
(WTF::FileSystemImpl::FileHandle::seek):
(WTF::FileSystemImpl::FileHandle::flush):
(WTF::FileSystemImpl::FileHandle::id):
* Source/WTF/wtf/FileHandle.h: Added.
(WTF::FileSystemImpl::FileHandle::FileHandle):
(WTF::FileSystemImpl::FileHandle::platformHandle const):
(WTF::FileSystemImpl::FileHandle::isValid const):
(WTF::FileSystemImpl::FileHandle::operator bool const):
* Source/WTF/wtf/FileSystem.cpp:
(WTF::FileSystemImpl::appendFileContentsToFileHandle):
(WTF::FileSystemImpl::MappedFileData::MappedFileData):
(WTF::FileSystemImpl::openAndLockFile):
(WTF::FileSystemImpl::createMappedFileData):
(WTF::FileSystemImpl::mapToFile):
(WTF::FileSystemImpl::readOrMakeSalt):
(WTF::FileSystemImpl::readEntireFile):
(WTF::FileSystemImpl::overwriteEntireFile):
(WTF::FileSystemImpl::createTemporaryFile):
(WTF::FileSystemImpl::unlockAndCloseFile): Deleted.
* Source/WTF/wtf/FileSystem.h:
(WTF::FileSystemImpl::openTemporaryFile):
* Source/WTF/wtf/cocoa/FileSystemCocoa.mm:
(WTF::FileSystemImpl::openTemporaryFile):
* Source/WTF/wtf/posix/FileSystemPOSIX.cpp:
(WTF::FileSystemImpl::openFile):
(WTF::FileSystemImpl::openTemporaryFile):
* Source/WTF/wtf/win/FileSystemWin.cpp:
(WTF::FileSystemImpl::openTemporaryFile):
(WTF::FileSystemImpl::openFile):
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/filesystemaccess/FileSystemStorageConnection.h:
(WebCore::FileSystemStorageConnection::SyncAccessHandleInfo::isolatedCopy):
* Source/WebCore/Modules/filesystemaccess/FileSystemSyncAccessHandle.cpp:
(WebCore::FileSystemSyncAccessHandle::create):
(WebCore::FileSystemSyncAccessHandle::FileSystemSyncAccessHandle):
(WebCore::FileSystemSyncAccessHandle::truncate):
(WebCore::FileSystemSyncAccessHandle::getSize):
(WebCore::FileSystemSyncAccessHandle::flush):
(WebCore::FileSystemSyncAccessHandle::read):
(WebCore::FileSystemSyncAccessHandle::write):
* Source/WebCore/Modules/filesystemaccess/FileSystemSyncAccessHandle.h:
* Source/WebCore/Modules/model-element/scenekit/SceneKitModelLoaderUSD.mm:
(WebCore::writeToTemporaryFile):
* Source/WebCore/Modules/webdatabase/OriginLock.cpp:
* Source/WebCore/Modules/webdatabase/OriginLock.h:
(): Deleted.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/GCController.cpp:
(WebCore::GCController::dumpHeapForVM):
* Source/WebCore/contentextensions/SerializedNFA.cpp:
(WebCore::ContentExtensions::writeAllToFile):
(WebCore::ContentExtensions::SerializedNFA::serialize):
* Source/WebCore/html/FileInputType.cpp:
(WebCore::FileInputType::filesChosen):
* Source/WebCore/loader/appcache/ApplicationCacheStorage.cpp:
(WebCore::ApplicationCacheStorage::writeDataToUniqueFileInDirectory):
* Source/WebCore/platform/FileHandle.cpp: Removed.
* Source/WebCore/platform/FileHandle.h: Removed.
* Source/WebCore/platform/FileStream.cpp:
(WebCore::FileStream::openForRead):
(WebCore::FileStream::close):
(WebCore::FileStream::read):
(WebCore::FileStream::FileStream): Deleted.
(WebCore::FileStream::~FileStream): Deleted.
* Source/WebCore/platform/FileStream.h:
* Source/WebCore/platform/cocoa/FileMonitorCocoa.mm:
(WebCore::FileMonitor::FileMonitor):
* Source/WebCore/platform/graphics/GraphicsLayer.cpp:
(showGraphicsLayerTree):
* Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
(WebCore::transcodeImage):
* Source/WebCore/platform/network/BlobRegistryImpl.cpp:
(WebCore::writeFilePathsOrDataBuffersToFile):
(WebCore::BlobRegistryImpl::writeBlobsToTemporaryFilesForIndexedDB):
* Source/WebCore/platform/network/curl/CookieJarDB.cpp:
(WebCore::CookieJarDB::flagDatabaseCorruption):
* Source/WebCore/platform/network/curl/CurlFormDataStream.cpp:
(WebCore::CurlFormDataStream::readFromFile):
* Source/WebCore/platform/network/curl/CurlFormDataStream.h:
* Source/WebCore/storage/StorageUtilities.cpp:
(WebCore::StorageUtilities::readOriginFromFile):
(WebCore::StorageUtilities::writeOriginToFile):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::createTemporaryFile):
* Source/WebCore/workers/service/server/SWScriptStorage.cpp:
(WebCore::SWScriptStorage::store):
* Source/WebGPU/WGSL/wgslc.cpp:
(runWGSL):
* Source/WebKit/NetworkProcess/NetworkDataTaskBlob.cpp:
(WebKit::NetworkDataTaskBlob::consumeData):
(WebKit::NetworkDataTaskBlob::download):
(WebKit::NetworkDataTaskBlob::writeDownload):
(WebKit::NetworkDataTaskBlob::cleanDownloadFiles):
(WebKit::NetworkDataTaskBlob::didFinishDownload):
(WebKit::NetworkDataTaskBlob::didFinish):
* Source/WebKit/NetworkProcess/NetworkDataTaskBlob.h:
* Source/WebKit/NetworkProcess/NetworkDataTaskDataURL.cpp:
(WebKit::NetworkDataTaskDataURL::downloadDecodedData):
* Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerDownloadTask.cpp:
(WebKit::ServiceWorkerDownloadTask::cancel):
(WebKit::ServiceWorkerDownloadTask::setPendingDownloadLocation):
(WebKit::ServiceWorkerDownloadTask::didReceiveData):
(WebKit::ServiceWorkerDownloadTask::didFinish):
(WebKit::ServiceWorkerDownloadTask::didFailDownload):
* Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerDownloadTask.h:
* Source/WebKit/NetworkProcess/cache/NetworkCache.cpp:
(WebKit::NetworkCache::Cache::dumpContentsToFile):
* Source/WebKit/NetworkProcess/cache/NetworkCacheData.cpp:
(WebKit::NetworkCache::Data::mapToFile const):
(WebKit::NetworkCache::mapFile):
(WebKit::NetworkCache::adoptAndMapFile):
* Source/WebKit/NetworkProcess/cache/NetworkCacheData.h:
(WebKit::NetworkCache::Data::Data):
* Source/WebKit/NetworkProcess/cache/NetworkCacheDataCocoa.mm:
(WebKit::NetworkCache::Data::adoptMap):
* Source/WebKit/NetworkProcess/cache/NetworkCacheDataCurl.cpp:
(WebKit::NetworkCache::Data::adoptMap):
* Source/WebKit/NetworkProcess/cache/NetworkCacheDataGLib.cpp:
(WebKit::NetworkCache::Data::Data):
(WebKit::NetworkCache::Data::adoptMap):
(WebKit::NetworkCache::Data::tryCreateSharedMemory const):
(WebKit::NetworkCache::MapWrapper::~MapWrapper): Deleted.
* Source/WebKit/NetworkProcess/cache/NetworkCacheIOChannel.h:
(WebKit::NetworkCache::IOChannel::WTF_GUARDED_BY_LOCK): Deleted.
* Source/WebKit/NetworkProcess/cache/NetworkCacheIOChannelCurl.cpp:
(WebKit::NetworkCache::IOChannel::~IOChannel):
(WebKit::NetworkCache::IOChannel::read):
(WebKit::NetworkCache::IOChannel::write):
* Source/WebKit/NetworkProcess/curl/NetworkDataTaskCurl.cpp:
(WebKit::NetworkDataTaskCurl::invokeDidReceiveResponse):
(WebKit::NetworkDataTaskCurl::deleteDownloadFile):
* Source/WebKit/NetworkProcess/curl/NetworkDataTaskCurl.h:
* Source/WebKit/NetworkProcess/storage/BackgroundFetchStoreManager.cpp:
(WebKit::BackgroundFetchStoreManager::storeFetchResponseBodyChunk):
* Source/WebKit/NetworkProcess/storage/CacheStorageManager.cpp:
(WebKit::readCachesList):
* Source/WebKit/NetworkProcess/storage/FileSystemStorageHandle.cpp:
(WebKit::FileSystemStorageHandle::create):
(WebKit::FileSystemStorageHandle::createSyncAccessHandle):
(WebKit::FileSystemStorageHandle::createWritable):
(WebKit::FileSystemStorageHandle::closeWritable):
(WebKit::FileSystemStorageHandle::executeCommandForWritableInternal):
(WebKit::FileSystemStorageHandle::computeCommandSpace):
* Source/WebKit/NetworkProcess/storage/FileSystemStorageHandle.h:
* Source/WebKit/Platform/IPC/SharedFileHandle.cpp:
(IPC::SharedFileHandle::create):
* Source/WebKit/Platform/IPC/SharedFileHandle.h:
(IPC::SharedFileHandle::release):
(IPC::SharedFileHandle::SharedFileHandle):
* Source/WebKit/Platform/IPC/cocoa/SharedFileHandleCocoa.cpp:
(IPC::SharedFileHandle::create):
(IPC::SharedFileHandle::SharedFileHandle):
(IPC::SharedFileHandle::toMachSendRight const):
* Source/WebKit/Shared/PersistencyUtils.cpp:
(WebKit::writeToDisk):
* Source/WebKit/Shared/WebMemorySampler.cpp:
(WebKit::WebMemorySampler::stop):
(WebKit::WebMemorySampler::initializeTempLogFile):
(WebKit::WebMemorySampler::writeHeaders):
(WebKit::WebMemorySampler::sampleTimerFired):
(WebKit::WebMemorySampler::appendCurrentMemoryUsageToFile):
* Source/WebKit/Shared/WebMemorySampler.h:
(): Deleted.
* Source/WebKit/Shared/mac/AuxiliaryProcessMac.mm:
(WebKit::fileContents):
(WebKit::writeSandboxDataToCacheFile):
* Source/WebKit/UIProcess/API/APIContentRuleListStore.cpp:
(API::writeDataToFile):
(API::compiledToFile):
(API::ContentRuleListStore::invalidateContentRuleListVersion):
(API::ContentRuleListStore::corruptContentRuleListHeader):
(API::ContentRuleListStore::invalidateContentRuleListHeader):
* Source/WebKit/UIProcess/Cocoa/SystemPreviewControllerCocoa.mm:
* Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
(WebKit::convertChromeExtensionToTemporaryZipFile):
* Source/WebKit/UIProcess/Inspector/win/WebInspectorUIProxyWin.cpp:
(WebKit::WebInspectorUIProxy::showSavePanelForSingleFile):
* Source/WebKit/UIProcess/WebFullScreenManagerProxy.cpp:
(WebKit::WebFullScreenManagerProxy::prepareQuickLookImageURL const):
* Source/WebKit/UIProcess/ios/WKModelView.mm:
(-[WKModelView createFileForModel:]):
* Source/WebKit/WebProcess/Model/mac/ARKitInlinePreviewModelPlayerMac.mm:
(WebKit::ARKitInlinePreviewModelPlayerMac::createFile):
* Tools/TestWebKitAPI/Tests/WTF/FileSystem.cpp:
(TestWebKitAPI::createTestFile):
(TestWebKitAPI::TEST_F(FileSystemTest, openExistingFileTruncate)):
(TestWebKitAPI::TEST_F(FileSystemTest, openExistingFileTruncateFailIfFileExists)):
(TestWebKitAPI::TEST_F(FileSystemTest, openExistingFileReadWrite)):
(TestWebKitAPI::TEST_F(FileSystemTest, openExistingFileReadWriteFailIfFileExists)):
(TestWebKitAPI::TEST_F(FileSystemTest, openExistingFileReadOnly)):
(TestWebKitAPI::TEST_F(FileSystemTest, openNonExistingFileTruncate)):
(TestWebKitAPI::TEST_F(FileSystemTest, openNonExistingFileTruncateFailIfFileExists)):
(TestWebKitAPI::TEST_F(FileSystemTest, openNonExistingFileReadWrite)):
(TestWebKitAPI::TEST_F(FileSystemTest, openNonExistingFileReadWriteFailIfFileExists)):
(TestWebKitAPI::TEST_F(FileSystemTest, openNonExistingFileReadOnly)):
(TestWebKitAPI::TEST_F(FileSystemTest, deleteEmptyDirectoryContainingDSStoreFile)):
(TestWebKitAPI::TEST_F(FileSystemTest, deleteEmptyDirectoryOnNonEmptyDirectory)):
(TestWebKitAPI::TEST_F(FileSystemTest, moveDirectory)):
(TestWebKitAPI::runGetFileModificationTimeTest):
(TestWebKitAPI::TEST_F(FileSystemTest, readEntireFile)):
* Tools/TestWebKitAPI/Tests/WebCore/FileMonitor.cpp:
(TestWebKitAPI::TEST_F(FileMonitorTest, DetectDeleteButNotSubsequentChange)):
* Tools/TestWebKitAPI/Tests/WebCore/SharedBufferTest.cpp:
(TestWebKitAPI::FragmentedSharedBufferTest::SetUp):
* Tools/TestWebKitAPI/Tests/WebCore/cocoa/DatabaseTrackerTest.mm:
(TestWebKitAPI::createFileAtPath):
* Tools/TestWebKitAPI/Tests/WebCore/gstreamer/GstElementHarness.cpp:
(TestWebKitAPI::TEST_F(GStreamerTest, harnessParseMP4)):
(TestWebKitAPI::TEST_F(GStreamerTest, harnessDecodeMP4Video)):

Canonical link: <a href="https://commits.webkit.org/292119@main">https://commits.webkit.org/292119@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/22a96ceeae3f1adcb31f20fbc9bda58d360833c6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/95037 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/14632 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/4489 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/100066 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/45536 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/97086 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/14921 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/23054 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/100066 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/45536 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/98038 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/11141 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/85812 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/100066 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/10852 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/3538 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/44875 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/87707 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/81043 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/3631 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/102111 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/93658 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/22079 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/16109 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/81499 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/22326 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/81830 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/80887 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20205 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/25445 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/2847 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/15321 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/22052 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/27175 "Build is in progress. Recent messages:") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/116348 "Built successfully") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/21709 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/33284 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/25184 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/23450 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

geoffreygaren commented on Source/WTF/wtf/FileHandle.cpp:
```diff
@@ -0,0 +1,160 @@
+namespace WTF::FileSystemImpl {
+
+FileHandle::FileHandle()
+    : m_handle(invalidPlatformFileHandle)
```
Can we use `std::optional` for `m_handle` to avoid in-band signaling here?

geoffreygaren reviewed (approved):
r=me

geoffreygaren commented on Source/WTF/wtf/FileHandle.cpp:
```diff
@@ -0,0 +1,158 @@
+    if (!isValid())
+        return;
+
+    auto handle = std::exchange(m_handle, std::nullopt).unsafeValue();
```
Since we're paying for filesystem operations anyway, we might as well use `value()` and drop the unsafety. Also, the `isValid()` check is trivially optimizable, so the compiler will just de-duplicate the code for us.

geoffreygaren commented on Source/WTF/wtf/FileHandle.h:
```diff
@@ -0,0 +1,115 @@
+
+    WTF_EXPORT_PRIVATE ~FileHandle();
+
+    PlatformFileHandle platformHandle() const { return m_handle.unsafeValue(); }
```
Since we're paying for filesystem operations anyway, we might as well use value() and drop the unsafety.

cdumez replied:
You want this?
```
PlatformFileHandle platformHandle() const { return m_handle ? *m_handle : invalidPlatformFileHandle; }
```

I can do that. It is identical to calling `unsafeValue()` in behavior, just with an extra branch so I personally felt it wasn't better. You're right there is no perf concern though.

cdumez replied:
Going with `m_handle.value_or(invalidPlatformFileHandle)` in this latest iteration. It looks "safer" though it is not any safer and it is a tiny bit less efficient.

cdumez replied:
I'll look in a follow-up since calling `value()` here causes assertions. Some call sites will need to be updated.

cdumez committed:
FileSystem::openFile() should return a FileHandle
https://bugs.webkit.org/show_bug.cgi?id=289509

Reviewed by Geoffrey Garen.

FileSystem::openFile() should return a FileHandle instead of a PlatformFileHandle, FileHandle being a
RAII wrapper to make sure we never forget to close file descriptors.

In a follow-up, I will look into cleaning up FileSystem.h, now that some of the API is only needed
by FileHandle.

* Source/JavaScriptCore/API/JSScript.mm:
(-[JSScript readCache]):
* Source/JavaScriptCore/jsc.cpp:
(JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/wasm/WasmStreamingParser.cpp:
(JSC::Wasm::dumpWasmSource):
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/FileHandle.cpp: Added.
(WTF::FileSystemImpl::FileHandle::FileHandle):
(WTF::FileSystemImpl::FileHandle::~FileHandle):
(WTF::FileSystemImpl::FileHandle::operator=):
(WTF::FileSystemImpl::FileHandle::readAll):
(WTF::FileSystemImpl::FileHandle::read):
(WTF::FileSystemImpl::FileHandle::write):
(WTF::FileSystemImpl::FileHandle::lock):
(WTF::FileSystemImpl::FileHandle::truncate):
(WTF::FileSystemImpl::FileHandle::closeIfNecessary):
(WTF::FileSystemImpl::FileHandle::size):
(WTF::FileSystemImpl::FileHandle::seek):
(WTF::FileSystemImpl::FileHandle::flush):
(WTF::FileSystemImpl::FileHandle::id):
* Source/WTF/wtf/FileHandle.h: Added.
(WTF::FileSystemImpl::FileHandle::FileHandle):
(WTF::FileSystemImpl::FileHandle::platformHandle const):
(WTF::FileSystemImpl::FileHandle::isValid const):
(WTF::FileSystemImpl::FileHandle::operator bool const):
* Source/WTF/wtf/FileSystem.cpp:
(WTF::FileSystemImpl::appendFileContentsToFileHandle):
(WTF::FileSystemImpl::MappedFileData::MappedFileData):
(WTF::FileSystemImpl::openAndLockFile):
(WTF::FileSystemImpl::createMappedFileData):
(WTF::FileSystemImpl::mapToFile):
(WTF::FileSystemImpl::readOrMakeSalt):
(WTF::FileSystemImpl::readEntireFile):
(WTF::FileSystemImpl::overwriteEntireFile):
(WTF::FileSystemImpl::createTemporaryFile):
(WTF::FileSystemImpl::unlockAndCloseFile): Deleted.
* Source/WTF/wtf/FileSystem.h:
(WTF::FileSystemImpl::openTemporaryFile):
* Source/WTF/wtf/cocoa/FileSystemCocoa.mm:
(WTF::FileSystemImpl::openTemporaryFile):
* Source/WTF/wtf/posix/FileSystemPOSIX.cpp:
(WTF::FileSystemImpl::openFile):
(WTF::FileSystemImpl::openTemporaryFile):
* Source/WTF/wtf/win/FileSystemWin.cpp:
(WTF::FileSystemImpl::openTemporaryFile):
(WTF::FileSystemImpl::openFile):
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/filesystemaccess/FileSystemStorageConnection.h:
(WebCore::FileSystemStorageConnection::SyncAccessHandleInfo::isolatedCopy):
* Source/WebCore/Modules/filesystemaccess/FileSystemSyncAccessHandle.cpp:
(WebCore::FileSystemSyncAccessHandle::create):
(WebCore::FileSystemSyncAccessHandle::FileSystemSyncAccessHandle):
(WebCore::FileSystemSyncAccessHandle::truncate):
(WebCore::FileSystemSyncAccessHandle::getSize):
(WebCore::FileSystemSyncAccessHandle::flush):
(WebCore::FileSystemSyncAccessHandle::read):
(WebCore::FileSystemSyncAccessHandle::write):
* Source/WebCore/Modules/filesystemaccess/FileSystemSyncAccessHandle.h:
* Source/WebCore/Modules/model-element/scenekit/SceneKitModelLoaderUSD.mm:
(WebCore::writeToTemporaryFile):
* Source/WebCore/Modules/webdatabase/OriginLock.cpp:
* Source/WebCore/Modules/webdatabase/OriginLock.h:
(): Deleted.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/GCController.cpp:
(WebCore::GCController::dumpHeapForVM):
* Source/WebCore/contentextensions/SerializedNFA.cpp:
(WebCore::ContentExtensions::writeAllToFile):
(WebCore::ContentExtensions::SerializedNFA::serialize):
* Source/WebCore/html/FileInputType.cpp:
(WebCore::FileInputType::filesChosen):
* Source/WebCore/loader/appcache/ApplicationCacheStorage.cpp:
(WebCore::ApplicationCacheStorage::writeDataToUniqueFileInDirectory):
* Source/WebCore/platform/FileHandle.cpp: Removed.
* Source/WebCore/platform/FileHandle.h: Removed.
* Source/WebCore/platform/FileStream.cpp:
(WebCore::FileStream::openForRead):
(WebCore::FileStream::close):
(WebCore::FileStream::read):
(WebCore::FileStream::FileStream): Deleted.
(WebCore::FileStream::~FileStream): Deleted.
* Source/WebCore/platform/FileStream.h:
* Source/WebCore/platform/cocoa/FileMonitorCocoa.mm:
(WebCore::FileMonitor::FileMonitor):
* Source/WebCore/platform/graphics/GraphicsLayer.cpp:
(showGraphicsLayerTree):
* Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
(WebCore::transcodeImage):
* Source/WebCore/platform/network/BlobRegistryImpl.cpp:
(WebCore::writeFilePathsOrDataBuffersToFile):
(WebCore::BlobRegistryImpl::writeBlobsToTemporaryFilesForIndexedDB):
* Source/WebCore/platform/network/curl/CookieJarDB.cpp:
(WebCore::CookieJarDB::flagDatabaseCorruption):
* Source/WebCore/platform/network/curl/CurlFormDataStream.cpp:
(WebCore::CurlFormDataStream::readFromFile):
* Source/WebCore/platform/network/curl/CurlFormDataStream.h:
* Source/WebCore/storage/StorageUtilities.cpp:
(WebCore::StorageUtilities::readOriginFromFile):
(WebCore::StorageUtilities::writeOriginToFile):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::createTemporaryFile):
* Source/WebCore/workers/service/server/SWScriptStorage.cpp:
(WebCore::SWScriptStorage::store):
* Source/WebGPU/WGSL/wgslc.cpp:
(runWGSL):
* Source/WebKit/NetworkProcess/NetworkDataTaskBlob.cpp:
(WebKit::NetworkDataTaskBlob::consumeData):
(WebKit::NetworkDataTaskBlob::download):
(WebKit::NetworkDataTaskBlob::writeDownload):
(WebKit::NetworkDataTaskBlob::cleanDownloadFiles):
(WebKit::NetworkDataTaskBlob::didFinishDownload):
(WebKit::NetworkDataTaskBlob::didFinish):
* Source/WebKit/NetworkProcess/NetworkDataTaskBlob.h:
* Source/WebKit/NetworkProcess/NetworkDataTaskDataURL.cpp:
(WebKit::NetworkDataTaskDataURL::downloadDecodedData):
* Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerDownloadTask.cpp:
(WebKit::ServiceWorkerDownloadTask::cancel):
(WebKit::ServiceWorkerDownloadTask::setPendingDownloadLocation):
(WebKit::ServiceWorkerDownloadTask::didReceiveData):
(WebKit::ServiceWorkerDownloadTask::didFinish):
(WebKit::ServiceWorkerDownloadTask::didFailDownload):
* Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerDownloadTask.h:
* Source/WebKit/NetworkProcess/cache/NetworkCache.cpp:
(WebKit::NetworkCache::Cache::dumpContentsToFile):
* Source/WebKit/NetworkProcess/cache/NetworkCacheData.cpp:
(WebKit::NetworkCache::Data::mapToFile const):
(WebKit::NetworkCache::mapFile):
(WebKit::NetworkCache::adoptAndMapFile):
* Source/WebKit/NetworkProcess/cache/NetworkCacheData.h:
(WebKit::NetworkCache::Data::Data):
* Source/WebKit/NetworkProcess/cache/NetworkCacheDataCocoa.mm:
(WebKit::NetworkCache::Data::adoptMap):
* Source/WebKit/NetworkProcess/cache/NetworkCacheDataCurl.cpp:
(WebKit::NetworkCache::Data::adoptMap):
* Source/WebKit/NetworkProcess/cache/NetworkCacheDataGLib.cpp:
(WebKit::NetworkCache::Data::Data):
(WebKit::NetworkCache::Data::adoptMap):
(WebKit::NetworkCache::Data::tryCreateSharedMemory const):
(WebKit::NetworkCache::MapWrapper::~MapWrapper): Deleted.
* Source/WebKit/NetworkProcess/cache/NetworkCacheIOChannel.h:
(WebKit::NetworkCache::IOChannel::WTF_GUARDED_BY_LOCK): Deleted.
* Source/WebKit/NetworkProcess/cache/NetworkCacheIOChannelCurl.cpp:
(WebKit::NetworkCache::IOChannel::~IOChannel):
(WebKit::NetworkCache::IOChannel::read):
(WebKit::NetworkCache::IOChannel::write):
* Source/WebKit/NetworkProcess/curl/NetworkDataTaskCurl.cpp:
(WebKit::NetworkDataTaskCurl::invokeDidReceiveResponse):
(WebKit::NetworkDataTaskCurl::deleteDownloadFile):
* Source/WebKit/NetworkProcess/curl/NetworkDataTaskCurl.h:
* Source/WebKit/NetworkProcess/storage/BackgroundFetchStoreManager.cpp:
(WebKit::BackgroundFetchStoreManager::storeFetchResponseBodyChunk):
* Source/WebKit/NetworkProcess/storage/CacheStorageManager.cpp:
(WebKit::readCachesList):
* Source/WebKit/NetworkProcess/storage/FileSystemStorageHandle.cpp:
(WebKit::FileSystemStorageHandle::create):
(WebKit::FileSystemStorageHandle::createSyncAccessHandle):
(WebKit::FileSystemStorageHandle::createWritable):
(WebKit::FileSystemStorageHandle::closeWritable):
(WebKit::FileSystemStorageHandle::executeCommandForWritableInternal):
(WebKit::FileSystemStorageHandle::computeCommandSpace):
* Source/WebKit/NetworkProcess/storage/FileSystemStorageHandle.h:
* Source/WebKit/Platform/IPC/SharedFileHandle.cpp:
(IPC::SharedFileHandle::create):
* Source/WebKit/Platform/IPC/SharedFileHandle.h:
(IPC::SharedFileHandle::release):
(IPC::SharedFileHandle::SharedFileHandle):
* Source/WebKit/Platform/IPC/cocoa/SharedFileHandleCocoa.cpp:
(IPC::SharedFileHandle::create):
(IPC::SharedFileHandle::SharedFileHandle):
(IPC::SharedFileHandle::toMachSendRight const):
* Source/WebKit/Shared/PersistencyUtils.cpp:
(WebKit::writeToDisk):
* Source/WebKit/Shared/WebMemorySampler.cpp:
(WebKit::WebMemorySampler::stop):
(WebKit::WebMemorySampler::initializeTempLogFile):
(WebKit::WebMemorySampler::writeHeaders):
(WebKit::WebMemorySampler::sampleTimerFired):
(WebKit::WebMemorySampler::appendCurrentMemoryUsageToFile):
* Source/WebKit/Shared/WebMemorySampler.h:
(): Deleted.
* Source/WebKit/Shared/mac/AuxiliaryProcessMac.mm:
(WebKit::fileContents):
(WebKit::writeSandboxDataToCacheFile):
* Source/WebKit/UIProcess/API/APIContentRuleListStore.cpp:
(API::writeDataToFile):
(API::compiledToFile):
(API::ContentRuleListStore::invalidateContentRuleListVersion):
(API::ContentRuleListStore::corruptContentRuleListHeader):
(API::ContentRuleListStore::invalidateContentRuleListHeader):
* Source/WebKit/UIProcess/Cocoa/SystemPreviewControllerCocoa.mm:
* Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
(WebKit::convertChromeExtensionToTemporaryZipFile):
* Source/WebKit/UIProcess/Inspector/win/WebInspectorUIProxyWin.cpp:
(WebKit::WebInspectorUIProxy::showSavePanelForSingleFile):
* Source/WebKit/UIProcess/WebFullScreenManagerProxy.cpp:
(WebKit::WebFullScreenManagerProxy::prepareQuickLookImageURL const):
* Source/WebKit/UIProcess/ios/WKModelView.mm:
(-[WKModelView createFileForModel:]):
* Source/WebKit/WebProcess/Model/mac/ARKitInlinePreviewModelPlayerMac.mm:
(WebKit::ARKitInlinePreviewModelPlayerMac::createFile):
* Tools/TestWebKitAPI/Tests/WTF/FileSystem.cpp:
(TestWebKitAPI::createTestFile):
(TestWebKitAPI::TEST_F(FileSystemTest, openExistingFileTruncate)):
(TestWebKitAPI::TEST_F(FileSystemTest, openExistingFileTruncateFailIfFileExists)):
(TestWebKitAPI::TEST_F(FileSystemTest, openExistingFileReadWrite)):
(TestWebKitAPI::TEST_F(FileSystemTest, openExistingFileReadWriteFailIfFileExists)):
(TestWebKitAPI::TEST_F(FileSystemTest, openExistingFileReadOnly)):
(TestWebKitAPI::TEST_F(FileSystemTest, openNonExistingFileTruncate)):
(TestWebKitAPI::TEST_F(FileSystemTest, openNonExistingFileTruncateFailIfFileExists)):
(TestWebKitAPI::TEST_F(FileSystemTest, openNonExistingFileReadWrite)):
(TestWebKitAPI::TEST_F(FileSystemTest, openNonExistingFileReadWriteFailIfFileExists)):
(TestWebKitAPI::TEST_F(FileSystemTest, openNonExistingFileReadOnly)):
(TestWebKitAPI::TEST_F(FileSystemTest, deleteEmptyDirectoryContainingDSStoreFile)):
(TestWebKitAPI::TEST_F(FileSystemTest, deleteEmptyDirectoryOnNonEmptyDirectory)):
(TestWebKitAPI::TEST_F(FileSystemTest, moveDirectory)):
(TestWebKitAPI::runGetFileModificationTimeTest):
(TestWebKitAPI::TEST_F(FileSystemTest, readEntireFile)):
* Tools/TestWebKitAPI/Tests/WebCore/FileMonitor.cpp:
(TestWebKitAPI::TEST_F(FileMonitorTest, DetectDeleteButNotSubsequentChange)):
* Tools/TestWebKitAPI/Tests/WebCore/SharedBufferTest.cpp:
(TestWebKitAPI::FragmentedSharedBufferTest::SetUp):
* Tools/TestWebKitAPI/Tests/WebCore/cocoa/DatabaseTrackerTest.mm:
(TestWebKitAPI::createFileAtPath):
* Tools/TestWebKitAPI/Tests/WebCore/gstreamer/GstElementHarness.cpp:
(TestWebKitAPI::TEST_F(GStreamerTest, harnessParseMP4)):
(TestWebKitAPI::TEST_F(GStreamerTest, harnessDecodeMP4Video)):

Canonical link: https://commits.webkit.org/292119@main
